Source/WebCore/platform/graphics/ImageFrameWorkQueue.cpp

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright (C) 2024 Apple Inc.  All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "ImageFrameWorkQueue.h"
+
+#include "ImageDecoder.h"
+#include "ImageSource.h"
+#include "Logging.h"
+#include <wtf/SystemTracing.h>
+
+namespace WebCore {
+
+Ref<ImageFrameWorkQueue> ImageFrameWorkQueue::create(ImageSource& source)
+{
+    return adoptRef(*new ImageFrameWorkQueue(source));
+}
+
+ImageFrameWorkQueue::ImageFrameWorkQueue(ImageSource& source)
+    : m_source(source)
+{
+}
+
+ImageFrameWorkQueue::RequestQueue& ImageFrameWorkQueue::requestQueue()
+{
+    if (!m_requestQueue)
+        m_requestQueue = RequestQueue::create();
+
+    return *m_requestQueue;
+}
+
+void ImageFrameWorkQueue::start()
+{
+    if (m_workQueue)
+        return;
+
+    ASSERT(isMainThread());
+    m_workQueue = WorkQueue::create("org.webkit.ImageDecoder", WorkQueue::QOS::Default);
+
+    m_workQueue->dispatch([protectedThis = Ref { *this }, protectedWorkQueue = Ref { *m_workQueue }, protectedSource = Ref { source() }] () mutable {
+        Request request;
+        while (protectedThis->requestQueue().dequeue(request)) {
```

Comment:
I need to look into the implementation of this requestQueue(). I assume it does locking?

Reply:
I see, it's a `SynchronizedFixedQueue` which does indeed Lock internally. All good.